demo/parser_demo.cpp

```diff
@@ -1,29 +1,71 @@
 #include <fstream>
 #include <iostream>
+#include "fmt/core.h"
+#include "fmt/color.h"
 #include "lexer.hpp"
 #include "parser.hpp"
 
 class PrintVisitor : public ast::Visitor {
+public:
+    PrintVisitor(size_t depth = 0) : depth(depth) {}
     void visit (ast::Program* node) override {
 
     }
     void visit(ast::Routine *node) override {
 
     };
     void visit(ast::Parameter *node) override {
-
+        if (node == nullptr) {
+            fmt::print("{:|>{}}- [Parameter]> null\n", "", depth);
+        } else {
+            fmt::print("{:|>{}}- [Parameter]> {}\n", "", depth, node->name.lit);
+            depth++;
+            visit(node->type.get());
+            depth--;
+        }
     };
     void visit(ast::Type *node) override {
```

Comment:
So you mean replacing line 23 with `node->type->accept(*this);` such that I don't need dynamic casting myself and the runtime will choose the appropriate overload automatically?

Reply:
I would also need to mark `Type::visit` as `virtual` (in ast.hpp, line 109), right?

Reply:
Look here https://stackoverflow.com/questions/11464735/visitor-pattern-for-ast 
Does this link answer?

Reply:
It answers the first part (that I need to use `accept` instead of `visit`), but there is no mention about the `accept` function in `Expression` and that it has to be `virtual`, but I assumed that it does have to be virtual